src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java

```diff
@@ -414,10 +422,18 @@ private void handleClientRoles(RoutingContext rc) {
             // upload to storage
             storeWriter.upload(clients, null);
 
+            ensurePublisherSiteKeyExists(c);
+
             // return client with new key
             rc.response().end(jsonWriter.writeValueAsString(c));
         } catch (Exception e) {
             rc.fail(500, e);
         }
     }
+
+    private void ensurePublisherSiteKeyExists(ClientKey clientKey) throws Exception {
+        if (clientKey.hasRole(Role.GENERATOR)) {
```

Comment:
actually Role.SHARER should also have site keys

Reply:
Leaving for now, but let's come back to it when we have settled on the new roles.